### PR TITLE
fix: add backend email field validation to prevent empty domains when allowing domain whitelist

### PR DESCRIPTION
**Pull request by karrui**
## Problem

Somehow even with frontend blocking, some forms are being created with `hasAllowedEmailDomains = true` but with empty `allowedEmailDomains`.

This PR fixes that by adding a backend schema validation on the email field to prevent such occurrences.

## Solution

**Fixes**:
- add email field validation to prevent empty domains when enabled

## Screenshots
_Note that this will not happen on prod, frontend validation was removed temporarily just to show validation being performed on the backend correctly._
Screenshot 2021-03-08 at 10 21 32 AM

## Tests
- [ ] Create email field with allowed email domains. Should save properly.
- [ ] For that same field, toggle allow email domains off. Should save properly. Go back into the field; the domains should have been cleared (in both client and network request object)
- [ ] Create email field with allow email domains toggled on. Client should prevent saving
  - [ ] when email domains given are empty
  - [ ] when duplicate email domains are given

### CODE REVIEW

File: src/app/models/field/emailField.ts [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
+const validateAllowDomainsToBeSet = (
+  emailDomains: string[],
+  hasAllowedEmailDomains: boolean,
+): boolean => {
+  // Case 1: hasAllowedEmailDomains is false and emailDomains must be empty.
+  const case1 = !hasAllowedEmailDomains && isEmpty(emailDomains)
+  // Case 2: hasAllowedEmailDomains is true and email domains must not be empty.
+  const case2 = hasAllowedEmailDomains && !isEmpty(emailDomains)
+
+  return case1 || case2
```
Comment (karrui) [inline src/app/models/field/emailField.ts:16]: Technically due to the pre-validate hook below, this can be simplified to

```ts
return !hasAllowedEmailDomains || !isEmpty(emailDomains)
```

but idk if I want to since that relies on the other ppl looking at this code making the connection

File: src/app/models/field/emailField.ts [hunk 2]
```diff
@@ -77,6 +105,15 @@ const createEmailFieldSchema = (): Schema<IEmailFieldSchema> => {
...
   })
 
+  // If hasAllowedEmailDomains is false, then clear allowedEmailDomains
+  EmailFieldSchema.pre<IEmailFieldSchema>('validate', function (next) {
+    if (!this.hasAllowedEmailDomains) {
+      this.allowedEmailDomains = []
+    }
+
+    return next()
+  })
```
Comment (liangyuanruo) [inline src/app/models/field/emailField.ts:115]: how about changing the database records to be consistent with the schema, instead of performing this lazy update?

Comment (karrui) [reply src/app/models/field/emailField.ts:115]: This can happen since the frontend gives the form object wholesale. An occurrence of both `hasAllowedEmailDomains = true` and a non-empty `allowedEmailDomains` can occur when:
1. current email field has email domains set
2. admin toggles allow email domain off (frontend doesn't clear already filled email domains)
3. this scenario gets sent to the backend

Instead of changing frontend code (which may still result in inconsistencies due to client-side manipulation), I'd thought doing it here is safer and more concrete

Comment (liangyuanruo) [reply src/app/models/field/emailField.ts:115]: > frontend doesn't clear already filled email domains ... Instead of changing frontend code (which may still result in inconsistencies due to client-side manipulation), I'd thought doing it here is safer and more concrete

This reasoning isn't correct - SPAs must always maintain a consistent viewmodel and still clear `allowedEmailDomains`.

File: src/app/models/field/emailField.ts (hunk 1 above)
Comment (liangyuanruo) [reply src/app/models/field/emailField.ts:16]: the existing code provides greater clarity, although I'd have used `emailDomains.length` instead of lodash's `isEmpty`

File: src/app/models/field/emailField.ts (hunk 2 above)
Comment (karrui) [reply src/app/models/field/emailField.ts:115]: Let me rephrase; could it be by design? A user may want to toggle it off, and toggle it back on to get their old results back before they submit? 

recording (17)

Comment (liangyuanruo) [reply src/app/models/field/emailField.ts:115]: is this a feature or a bug? The inconsistency is also the result of the model being overly flexible, we shouldn't have introduced the `hasAllowedEmailDomains` boolean - instead, it should have only been computed on the fly by the frontend.

Comment (liangyuanruo) [reply src/app/models/field/emailField.ts:115]: in the meantime, why don't we clear the array when the toggle is touched? I don't think the UX change is going to inconvenience users.

Comment (karrui) [reply src/app/models/field/emailField.ts:115]: Will do

Comment (liangyuanruo) [reply src/app/models/field/emailField.ts:115]: FYI I created another [issue](https://github.com/opengovsg/FormSG/issues/1312) to track the inconsistency problem

Comment (karrui) [reply src/app/models/field/emailField.ts]: Cleared array when toggled off in d84eac1, and removed the pre-validate transformation in d665caa

**Review by liangyuanruo — CHANGES_REQUESTED** [review]
As discussed, we will drop `hasAllowedEmailDomains` from the schema and address https://github.com/opengovsg/FormSG/issues/1312 directly.

**Comment by karrui** [comment]
Closed until updateField API refactor is done
